src/promql/src/range_array.rs

```diff
@@ -0,0 +1,365 @@
+    }
+
+    pub fn from_ranges<R>(values: ArrayRef, ranges: R) -> Result<Self>
+    where
+        R: IntoIterator<Item = (u32, u32)> + Clone,
+    {
+        Self::check_ranges(values.len(), ranges.clone())?;
+
+        unsafe { Ok(Self::from_ranges_unchecked(values, ranges)) }
+    }
+
+    /// Construct [RangeArray] from given range without checking its validaty.
+    ///
+    /// # Safety
+    ///
+    /// Caller should ensure the given range are valid. Otherwise use [`from_ranges`]
+    /// instead.
+    ///
+    /// [`from_ranges`]: crate::range_array::RangeArray#method.from_ranges
+    pub unsafe fn from_ranges_unchecked<R>(values: ArrayRef, ranges: R) -> Self
+    where
+        R: IntoIterator<Item = (u32, u32)>,
+    {
+        let key_array = Int64Array::from_iter(
+            ranges
+                .into_iter()
+                .map(|(offset, length)| pack(offset, length)),
+        );
+
+        // Build from ArrayData to bypass the "offset" checker. Because
+        // we are not using "keys" as-is.
+        // This paragraph is copied from arrow-rs dictionary_array.rs `try_new()`.
+        let mut data = ArrayData::builder(DataType::Dictionary(
+            Box::new(Self::key_type()),
+            Box::new(values.data_type().clone()),
+        ))
+        .len(key_array.len())
+        .add_buffer(key_array.data().buffers()[0].clone())
+        .add_child_data(values.data().clone());
+        match key_array.data().null_buffer() {
+            Some(buffer) if key_array.data().null_count() > 0 => {
+                data = data
+                    .null_bit_buffer(Some(buffer.clone()))
+                    .null_count(key_array.data().null_count());
+            }
+            _ => data = data.null_count(0),
+        }
+        let array_data = unsafe { data.build_unchecked() };
+
+        Self {
+            array: array_data.into(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.array.keys().len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.array.keys().is_empty()
+    }
+
+    pub fn get(&self, index: usize) -> Option<ArrayRef> {
+        if index >= self.len() {
+            return None;
+        }
+
+        let compound_key = self.array.keys().value(index);
+        let (offset, length) = unpack(compound_key);
+        let array = self.array.values().slice(offset as usize, length as usize);
+
+        Some(array)
+    }
+
+    /// Return the underlying Arrow's [DictionaryArray]. Notes the dictionary array might be
+    /// invalid from Arrow's definition. Be care if try to access this dictionary through
+    /// Arrow's API.
+    pub fn into_dict(self) -> DictionaryArray<Int64Type> {
+        self.array
+    }
+
+    fn check_ranges<R>(value_len: usize, ranges: R) -> Result<()>
+    where
+        R: IntoIterator<Item = (u32, u32)>,
+    {
+        for (offset, length) in ranges.into_iter() {
+            ensure!(
+                offset + length <= value_len as _,
```

Comment:
Better to cast offset and length to `usize` like line 150, after all `(u32::MAX, 1)` could be a valid range.
```suggestion
                offset as usize + length as usize <= value_len,
```

Reply:
👍